### PR TITLE
Add a new challenge to verify `CStr`

### PR DESCRIPTION
**Pull request by celinval**
Proposes a new challenge to verify the safety of `CStr` implementation.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: doc/src/challenges/0013-cstr.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+## Description
+
+The goal of this challenge is to ensure the safety of the CStr struct implementation.
+First, we need to specify a safety invariant that captures the essential safety properties that must be maintained.
+
+Next, we should verify that all the safe, public methods respect this invariant.
+For example, we can check that creating a `CStr` from a byte slice with method `from_bytes_with_nul` will only yield
+safe values of `CStr`.
+
+Finally, for unsafe methods like `as_ptr()` and `from_ptr()`, we need to specify appropriate safety contracts.
```
Comment (zhassan-aws) [inline doc/src/challenges/0013-cstr.md:33]: `as_ptr` is a safe method. Did you mean something else?

File: doc/src/challenges/0013-cstr.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+The `CStr` structure is meant to be used to build safe wrappers of FFI functions that may leverage `CStr::as_ptr`
+and the unsafe `CStr::from_ptr` constructor to provide a safe interface to other consumers.
+It provides safe methods to convert `CStr` to a Rust `&str` by performing UTF-8 validation, or into an owned `CString`.
+
+Any issue with this structure or misusage of its unsafe methods could trigger an invalid memory access, which poses
+a security risk for their users.
+
+## Description
+
+The goal of this challenge is to ensure the safety of the CStr struct implementation.
```
Comment (zhassan-aws) [inline doc/src/challenges/0013-cstr.md:26]:
```suggestion
The goal of this challenge is to ensure the safety of the `CStr` struct implementation.
```

File: doc/src/challenges/0013-cstr.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+### Assumptions
+
+- Harnesses may be bounded.
+
+### Success Criteria
+
+1. Implement the `Invariant` trait for `CStr`.
+
+2. Verify that the `CStr` safety invariant holds after calling any of the following public safe methods.
```
Comment (zhassan-aws) [inline doc/src/challenges/0013-cstr.md:45]: Shouldn't it hold after calling the public unsafe methods as well?

**Review by zhassan-aws — APPROVED** [review]
Thanks!

File: doc/src/challenges/0013-cstr.md (hunk 1 above)
Comment (celinval) [reply doc/src/challenges/0013-cstr.md:33]: oops... let me fix that

File: doc/src/challenges/0013-cstr.md (hunk 3 above)
Comment (celinval) [reply doc/src/challenges/0013-cstr.md:45]: yes, I mention that above: 
> These contracts should ensure no undefined behavior occurs within the unsafe methods themselves,
and that they maintain the overall safety invariant of `CStr` when called correctly.

File: doc/src/challenges/0013-cstr.md (hunk 1 above)
Comment (celinval) [reply doc/src/challenges/0013-cstr.md:33]:
```suggestion
Finally, for unsafe methods like `from_ptr()` and `from_bytes_with_nul_unchecked`, we need to specify appropriate safety contracts.
```

**Comment by celinval** [comment]
Oh, I forgot to add the trait implementations for this struct. Let me add them in the next revision.

File: doc/src/challenges/0013-cstr.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+| Function               | Location           |
```
Comment (qinheping) [inline doc/src/challenges/0013-cstr.md:47]: There is another public safe method `as_ptr`, why isn't it in the list?

**Review by qinheping — APPROVED** [review]
LGTM, thank you!

File: doc/src/challenges/0013-cstr.md (hunk 4 above)
Comment (celinval) [reply doc/src/challenges/0013-cstr.md:47]: Sure, I can add it. I was trying to add things that either performed some unsafe operation, modified or produced an object.

**Review by celinval — COMMENTED** [review]
trigger review check

**Pull request merged**